torchtext/data/metrics.py

```diff
@@ -0,0 +1,93 @@
+            tokens
+        references_corpus: a list of lists of reference translations. Each translation
+            is a list of tokens
+        max_n: the maximum n-gram we want to use. E.g. if max_n=3, we will use unigrams,
+            bigrams and trigrams
+        weights: weights used for each n-gram category (uniform by default)
+
+    Examples:
+        >>> candidate_corpus = [['I', 'ate', 'the', 'apple'], ['I', 'did']]
+        >>> references_corpus = [[['I', 'ate', 'it'], ['I', 'ate', 'apples']],
+                [['I', 'did']]]
+        >>> bleu_score(candidate_corpus, references_corpus)
+        >>> 0.7598356856515925
+    """
+
+    def _compute_ngram_counter(tokens, max_n):
+        """ Create a Counter with a count of unique n-grams in the tokens list
+
+        Arguments:
+            tokens: a list of tokens (typically a string split on whitespaces)
+            max_n: the maximum order of n-gram wanted
+
+        Outputs:
+            output: a collections.Counter object with the unique n-grams and their
+                associated count
+
+        Examples:
+            >>> from torchtext.data.functional import _compute_ngram_counter
+            >>> tokens = ['me', 'me', 'you']
+            >>> _compute_ngram_counter(tokens, 2)
+                Counter({('me',): 2,
+                 ('you',): 1,
+                 ('me', 'me'): 1,
+                 ('me', 'you'): 1,
+                 ('me', 'me', 'you'): 1})
+        """
+        assert max_n > 0
+
+        ngrams = [tuple(x.split(' ')) for x in ngrams_iterator(tokens, max_n)]
+        ngrams_counter = collections.Counter(ngrams)
+
+        return ngrams_counter
+
+    assert max_n == len(weights), 'Length of the "weights" list has be equal to max_n'
+
+    clipped_counts = [0.0] * max_n
+    total_counts = [0.0] * max_n
+
+    candidate_len = 0.0
+    refs_len = 0.0
+
+    for (candidate, refs) in zip(candidate_corpus, references_corpus):
+        candidate_len += len(candidate)
+
+        # Get the length of the reference that's closest in length to the candidate
+        refs_len_list = [float(len(ref)) for ref in refs]
+        refs_len = min(refs_len_list, key=lambda x: abs(len(candidate) - x))
+
+        reference_counters = collections.Counter()
+        for ref in refs:
+            reference_counters = reference_counters | _compute_ngram_counter(ref, max_n)
+
+        candidate_counter = _compute_ngram_counter(candidate, max_n)
+
+        clipped_counter = candidate_counter & reference_counters
+
+        for ngram in clipped_counter:
+            clipped_counts[len(ngram) - 1] += clipped_counter[ngram]
+
+        for ngram in candidate_counter:  # TODO: no need to loop through the whole counter
+            total_counts[len(ngram) - 1] += candidate_counter[ngram]
+
+    if min(clipped_counts) == 0:
+        return 0.0
+    else:
+        pn = [clipped_counts[i] / total_counts[i] for i in range(max_n)]
```

Comment:
A lot of the math here can be done using torch.Tensors - since this is part of the DAPI ecosystem it's ok to hard a hard dependency on that. Numpy on the other hand is something we can't have a hard dependency on, because pytorch/pytorch doesn't either.

Reply:
good point, I'll look into it

Reply:
I replaced the math part with tensor operations in 4fc19dc8524bff9d9fe8b314645eda8bf547e9eb

Makes it more elegant without the list comprehensions!